src/vos/tests/vts_mvcc.c

```diff
@@ -183,25 +196,87 @@ set_value(int i, char *path, d_iov_t *value)
 	value->iov_len = strlen(value->iov_buf) + 1;
 }
 
+static struct dtx_handle *
+start_tx(daos_handle_t coh, daos_unit_oid_t oid, daos_epoch_t epoch,
+	 struct tx_helper *txh)
+{
+	struct dtx_handle	*dth;
+
+	if (txh == NULL)
+		return NULL;
+
+	dth = txh->th_dth;
+
+	if (dth == NULL) {
+		vts_dtx_begin_ex(&oid, coh, epoch, 0, txh->th_nr_mods,
+				 &dth);
+		txh->th_dth = dth;
+	}
+
+	return dth;
+}
+
+static void
+stop_tx(daos_handle_t coh, struct tx_helper *txh, bool success, bool write)
+{
+	struct dtx_handle	*dth = txh->th_dth;
+	struct dtx_id		 xid;
+	bool			 has_committed = txh->th_has_committed;
+	int			 err;
+
+	if (txh == NULL)
+		return;
+
+	/** Mark if PMDK transaction would have committed so we know to abort */
+	if (write) {
+		if (success && dth->dth_modification_cnt &&
+		    dth->dth_op_seq == dth->dth_modification_cnt)
+			has_committed = true;
+		dth->dth_op_seq++;
+	}
+
+	if (txh->th_nr_ops == txh->th_op_seq) {
+		xid = dth->dth_xid;
+		vts_dtx_end(dth);
+		if (txh->th_nr_mods != 0) {
+			if (success) {
+				err = vos_dtx_commit(coh, &xid, 1, NULL);
+				assert_int_equal(err, 1);
+			} else {
+				if (txh->th_has_committed) {
```

Comment:
"th_has_committed" or "!th_has_committed" ?

Reply:
I removed this block.   It was never executed.